Documentation panel redesign

vitvakatu opened the pull request:
### Pull Request Description

Based on #7350 

Closes #7356 

- Documentation parser now supports two new DocSection types: 
  -  List is a better representation of unordered lists anywhere in the text.
  - Arguments, a list starting with the `Arguments:` keyword. Each item in the list is split into argument names and descriptions, and we use this information to highlight argument names in the documentation.
   I'm not confident with my changes to the doc parser, especially due to the fact there were exactly zero tests for it.
- Small adjustments to the documentation CSS. It now looks much closer to the design, though not ideal. In particular, paddings are different now, and no topmost headers are displayed. (as they are replaced with the breadcrumbs in the new design) I also used fixed sizes like `4px` instead of `0.25rem` to simplify matching with the design. I believe it is the way as we want the most accurate translation from Figma, even though it is generally not the best practice.
- Improvements for tags styles. The proper implementation (with different colors, limited count, custom per-tag formatting, `...` button...) will require more effort.
- Breadcrumbs are in *almost* proper position now. The exact positioning should be done after updating them to the new design.
- Added a missed support for `code` tags in the documentation. Now text emphasized with backticks has a separate background color.

https://github.com/enso-org/enso/assets/6566674/bc53d94e-c958-4bda-920d-70a1a8cfcb4e


### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

vitvakatu committed:
Support arguments list in the doc parser

vitvakatu committed:
Support new doc sections in documentation panel

vitvakatu committed:
Remove headers

vitvakatu committed:
Remove outer dom and place breadcrumbs

vitvakatu committed:
Fix methods icon

vitvakatu committed:
Use unordered list class in css

vitvakatu committed:
Improve tags styles

vitvakatu committed:
Remove virtual component groups docs

vitvakatu committed:
Cleanup top-level css styles

vitvakatu committed:
Small adjustments to headers

vitvakatu committed:
Add styles for emphasized text

vitvakatu committed:
Add bold font for arguments

vitvakatu committed:
Self-review

farmaazon commented on app/gui/suggestion-database/src/documentation_ir.rs:
```diff
@@ -286,13 +286,11 @@ impl EntryDocumentation {
 /// No documentation is available for the entry.
-#[derive(Debug, Clone, CloneRef, PartialEq)]
+#[derive(Debug, Clone, Copy, CloneRef, PartialEq)]
 #[allow(missing_docs)]
 pub enum Placeholder {
     /// Documentation is empty.
     NoDocumentation,
-    /// Documentation for the Virtual Component group.
-    VirtualComponentGroup { name: ImString },
 }
```
I guess we could make it a structure? Or it's not needed anymore?

MichaelMauderer replied:
It's not needed anymore and I removed it.

kazcw commented on app/gui/view/documentation/src/lib.rs:
```diff
@@ -155,24 +154,14 @@ impl Model {
-        self.breadcrumbs.frp().set_size(Vector2(size.x, 32.0));
-    }
-
     /// Display the documentation and scroll to the qualified name if needed.
```
This comment seems outdated (already before this change)

kazcw commented on app/gui/view/documentation/src/lib.rs:
```diff
@@ -200,16 +189,25 @@ impl Model {
+        self.dom.set_style_or_warn("border-radius", format!("{}px", style.corner_radius));
+
+        self.breadcrumbs.set_xy(Vector2(0.0, size.y));
+        self.breadcrumbs.frp().set_size(Vector2(size.x, style.breadcrumbs_height));
```
I don't understand how these lines are organized--it is partially by the component, but some of the `corner_radius` configuration has its own section?

MichaelMauderer replied:
I've moved them to be organised by component.

kazcw commented on lib/rust/parser/doc-parser/src/doc_sections.rs:
```diff
@@ -155,6 +186,9 @@ impl<L> TokenConsumer<L> for DocSectionCollector {
     fn enter_keyed_section(&mut self, header: Span<'_, L>) {
         self.finish_section();
         let key = header.to_string();
+        if key.to_lowercase() == "arguments" {
```
```suggestion
        if key.eq_ignore_ascii_case("Arguments") {
```
We process a lot of documentation; it's important to avoid allocations when we can.

kazcw commented on lib/rust/parser/doc-parser/src/doc_sections.rs:
```diff
@@ -118,22 +144,27 @@ impl DocSectionCollector {
             | Some(DocSection::Marked { body, .. }) => *body = text,
-            Some(DocSection::Tag { .. }) | None =>
-                self.sections.push(DocSection::Paragraph { body: text }),
+            Some(DocSection::Tag { .. })
+            | Some(DocSection::List { .. })
+            | Some(DocSection::Arguments { .. })
+            | None => self.sections.push(DocSection::Paragraph { body: text }),
```
This is causing the empty `Paragraph` between the `List` and the "Example:" section in the new test. The empty paragraph is not needed; the newlines between sections are implicit.

```suggestion
            Some(DocSection::List { .. }) | Some(DocSection::Arguments { .. }) => (),
            Some(DocSection::Tag { .. }) | None => self.sections.push(DocSection::Paragraph { body: text }),
```

kazcw commented on lib/rust/parser/doc-parser/src/doc_sections.rs:
```diff
@@ -203,8 +235,19 @@ impl<L> TokenConsumer<L> for DocSectionCollector {
-            ScopeType::List => self.current_body.push_str("</ul>"),
-            ScopeType::ListItem => (),
+            ScopeType::List => {
+                let items = mem::take(&mut self.current_list);
```
This gives away our buffer. If we `take` it, we need `log(size)` allocations per list. If we iterate/clone it and then `clear` it, we only need one allocation per `List`.

kazcw commented on lib/rust/parser/doc-parser/src/doc_sections.rs:
```diff
@@ -203,8 +235,19 @@ impl<L> TokenConsumer<L> for DocSectionCollector {
+                }
+            }
+            ScopeType::ListItem => {
+                self.current_list.push(mem::take(&mut self.current_body));
```
```suggestion
                self.current_list.push(self.current_body.drain(..).collect());
```

kazcw commented on lib/rust/parser/doc-parser/src/lib.rs:
```diff
@@ -688,3 +689,61 @@ pub trait TokenConsumer<L> {
+                }].to_vec()
+            },
+            List { items: ["List item 1".into(), "List item 2".into(), "List item 3".into()].to_vec() },
+            Paragraph { body: "".into() },
```
Spurious, see above.

somebody1234 reviewed (approved):
it's the same (automatic) `package-lock.json` change as in some other PRs.
when merging with develop, this change should go away, because said change has been merged into `develop`:

https://github.com/enso-org/enso/blob/73237b7b2552ac81b2db34d052812d41556396da/app/ide-desktop/package-lock.json#L447

MichaelMauderer committed:
Remove redundant placeholder struct.

MichaelMauderer committed:
Update outdated doc.

MichaelMauderer committed:
Avoid allocation when comparing strings.

MichaelMauderer committed:
Avoid empty paragraph.

MichaelMauderer committed:
Reduce allocations.

MichaelMauderer committed:
Update test to remove empty paragraph.

MichaelMauderer committed:
Fix rebase issues.

MichaelMauderer committed:
Improve padding and size handling in UI themes

Added padding_x and padding_y to hardcoded theme's breadcrumb settings to ensure consistent padding. Also, these padding settings and breadcrumb_height are now used directly in the Style structure, eliminating hardcoded values in the view documentation.

MichaelMauderer committed:
Adjusted breadcrumb background dimensions calculation.

Frizi commented:
QA Status: 🔴 
---

Overall it looks good, but there are still some minor rendering issues in some edge cases. It looks like some of them are caused by the documentation not having a consistent syntax in a few places.

1. Some methods have truncated argument lists. The documentation is not properly formatted, such that the argument list is missing a colon after the argument name. Though I think that it should be displayed in some way anyway. Instead, now is displayed as a list with empty items.
<img width="506" alt="Zrzut ekranu 2023-08-08 141610" src="https://github.com/enso-org/enso/assets/919491/739b07c7-fe5f-45c1-80f5-b5bcbe7b074a">

2. We are currently displaying all tags. I don't think that `ICON` tag should really be shown. We already display the icon itself after all. Though this is not a big issue and might be addressed later if needed.
<img width="316" alt="Zrzut ekranu 2023-08-08 141725" src="https://github.com/enso-org/enso/assets/919491/7a19c8c7-b00a-414a-a5f5-c05458957b7d">

3. The selection behaves strangely right after typing. There are two distinct selection indicators (do we still need that?), and they point to different entries. Additionally, the list is scrolled up, such that the actual selected entry is partially outside the view.

EDIT: I think this was a separate issue that is already fixed on develop (hence I have only noticed it in this branch). Updating the branch should resolve it.
 
https://github.com/enso-org/enso/assets/919491/9d072afe-9d80-4b7a-8cfb-d90f875b2403

vitvakatu commented:
I'll take it from here, thanks for testing.

vitvakatu committed:
Add support for improper arguments formatting in documenation comments

vitvakatu committed:
Do not include Icon tag into the docs

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/doc-panel-redesign

vitvakatu commented:
@Frizi I fixed both issues.
1. Now we correctly parse space-separated arguments.
2. `Icon` tag is now ignored in the documentation.

Could you retest?

xvcgreg commented:
@Frizi can you retest it?

Frizi commented:
QA Status: 🟡 
----
The above issues has been fixed, but i've noticed that invalid behavior of the documentation show/hide action has been reintroduced. The content of the panel should not be reflowed on every frame of the animation. Instead, it should be clipped to fit the animated background size, without acutally changing the content width of underlying DOM element.

See point 2. in https://github.com/enso-org/enso/pull/7350#issuecomment-1653649692

No other issues spotted, so once show/hide is fixed it's good to go.

vitvakatu committed:
Fix documentation panel resizing

vitvakatu committed:
enso-formatter

vitvakatu committed:
Merge branch 'develop' into wip/vitvakatu/doc-panel-redesign
